components/mpas-ocean/src/shared/mpas_ocn_time_average_coupled.F

```diff
@@ -584,4 +549,129 @@ subroutine ocn_time_average_coupled_accumulate(statePool, forcingPool, timeLevel
 
     end subroutine ocn_time_average_coupled_accumulate!}}}
 
+
+!***********************************************************************
+!
+!  routine accumulate_glc_3d_thermal_forcing
+!
+!> \brief   Coupled time averager accumulation for 3D thermal forcing
+!> \author  Xylar Asay-Davis
+!> \date    03/05/2025
+!> \details
+!>  This routine accumulated the coupled time average of the 3D thermal
+!>  forcing
+!
+!-----------------------------------------------------------------------
+    subroutine accumulate_glc_3d_thermal_forcing(statePool, forcingPool, timeLevel)!{{{
+        use ocn_constants, only: &
+             latent_heat_fusion_mks
+
+        type (mpas_pool_type), intent(in) :: statePool
+        type (mpas_pool_type), intent(inout) :: forcingPool
+        integer, intent(in) :: timeLevel
+
+        integer :: iCell
+        real (kind=RKIND), dimension(:,:), pointer :: avgThermalForcingAtZLevels
+        real (kind=RKIND), dimension(:,:), pointer :: avgThermalForcingAtZLevelsMask
+        real (kind=RKIND), dimension(:), pointer :: glcZLevels
+        real (kind=RKIND), dimension(:), pointer :: ssh
+
+        type (mpas_pool_type), pointer :: tracersPool
+
+        real (kind=RKIND), dimension(:,:,:), pointer :: activeTracers
+        integer, pointer :: indexTemperature, indexSalinity, nGlcZLevels, &
+                            nCells, nAccumulatedCoupled
+        integer :: iLevel, iLevelGlc, iLevelAbove, iLevelBelow
+
+        real (kind=RKIND) :: freezingTemp, glcz, zMin, zMax, zAbove, zBelow, &
+                             alpha, salin, temp, press
+
+        call mpas_pool_get_dimension(forcingPool, 'nCells', nCells)
+        call mpas_pool_get_dimension(forcingPool, 'nGlcZLevels', nGlcZLevels)
+
+        call mpas_pool_get_array(forcingPool, 'nAccumulatedCoupled', nAccumulatedCoupled)
+
+        call mpas_pool_get_array(forcingPool, 'glcZLevels', glcZLevels)
+        call mpas_pool_get_array(statePool, 'ssh', ssh, timeLevel)
+
+        call mpas_pool_get_array(forcingPool, 'avgThermalForcingAtZLevels', avgThermalForcingAtZLevels)
+        call mpas_pool_get_array(forcingPool, 'avgThermalForcingAtZLevelsMask', avgThermalForcingAtZLevelsMask)
+        call mpas_pool_get_subpool(statePool, 'tracers', tracersPool)
+        call mpas_pool_get_array(tracersPool, 'activeTracers', activeTracers, timeLevel)
+        call mpas_pool_get_dimension(tracersPool, 'index_temperature', indexTemperature)
+        call mpas_pool_get_dimension(tracersPool, 'index_salinity', indexSalinity)
+
+        !$omp parallel
+        !$omp do schedule(runtime)
+        do iLevelGlc = 1, nGlcZLevels
+            ! calculate thermal forcing at identified level for each cell
+            glcz = glcZLevels(iLevelGlc)
+
+            do iCell = 1, nCells
+
+                if (-glcz > ssh(iCell) .or. glcz > bottomDepth(iCell)) then
```

Comment:
@matthewhoffman, are you sure about this change?  It looked like `zocn_levels` was positive-up (i.e. negative).  If `glcz` is positive down, this is not the only place this needs to be fixed.

Reply:
@xylar , `zocn_levels` is positive-up in the coupler and in MALI, but when I imported it into MPAS-Ocean, I flipped the sign so it would be a z-level in the sense that MPAS-Ocean defines its vertical coordinate (line 3426 in `components/mpas-ocean/driver/ocn_comp_mct.F`):
https://github.com/E3SM-Ocean-Discussion/E3SM/pull/121/files#diff-ad4cc22e8040873be0cfd9d39fb6587d009034734e4e8c5a580f68acc644c765R3426
` glcZLevels = -1.0_RKIND * glc_get_zlevels()`

If you think that is too confusing, we can eliminate the sign flip and revert that last commit.  (Or keep it and fix the other places that would need adjusting.)

Reply:
Yes, I think that's quite confusing because MPAS-Ocean also uses positive-up z coordinates.

Reply:
`refBottomDepth` and `bottomDepth` are unfortunate exceptions but `zMid`, `ssh`, etc. are defined as positive-up.